src/client/api/metrics.c

```diff
@@ -0,0 +1,914 @@
+#include <daos/object.h>
+
+/** Tunable to enable/disable metrics */
+int is_metrics_enabled = 0;
+
+/** Linked list for TLS metric data */
+static D_LIST_HEAD(metrics_list);
+static pthread_mutex_t metrics_tls_lock = PTHREAD_MUTEX_INITIALIZER;
+
+/** Thread specific tls data pointer */
+__thread dc_metrics_tls_data_t *metrics_tls;
+pthread_key_t metrics_tls_key;
+
+/** Holds the aggregated values of metrics from terminated threads */
+static dc_metrics_tls_data_t metrics_tls_agg;
+
+#define	RPDIST_ENUM_IDX_CNT	129
+#define	ECDIST_ENUM_IDX_CNT	17
+static int rpdist_enum_idx[RPDIST_ENUM_IDX_CNT];
+static int ecdist_enum_idx[ECDIST_ENUM_IDX_CNT];
+
+static void dc_metrics_tls_cleanup(void *value);
+
+int
+dc_metrics_tls_init()
+{
+	int rc = 0;
+	rc = pthread_key_create(&metrics_tls_key, dc_metrics_tls_cleanup);
+	if (rc) {
+		D_GOTO(out, rc = -DER_NOMEM);
+	}
+	metrics_tls_agg.update_stat.st_min = ULONG_MAX;
+	metrics_tls_agg.fetch_stat.st_min = ULONG_MAX;
+	D_MUTEX_LOCK(&metrics_tls_lock);
+	d_list_add(&metrics_tls_agg.list, &metrics_list);
+	D_MUTEX_UNLOCK(&metrics_tls_lock);
+out:
+	return rc;
+}
+
+int
+dc_metrics_tls_alloc()
+{
+	int rc = 0;
+
+	if (metrics_tls != NULL) {
+		goto out;
+	}
+	D_ALLOC_PTR(metrics_tls);
+	if (metrics_tls == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
+	metrics_tls->update_stat.st_min = ULONG_MAX;
+	metrics_tls->fetch_stat.st_min = ULONG_MAX;
+	D_MUTEX_LOCK(&metrics_tls_lock);
+	d_list_add(&metrics_tls->list, &metrics_list);
+	pthread_setspecific(metrics_tls_key, metrics_tls);
+	D_MUTEX_UNLOCK(&metrics_tls_lock);
+out:
+	return rc;
+}
+
+static void
+dc_metrics_reset_tls_data()
+{
+	dc_metrics_tls_data_t *entry;
+
+	D_MUTEX_LOCK(&metrics_tls_lock);
+	d_list_for_each_entry(entry, &metrics_list, list) {
+		memset(&entry->update_stat, 0, sizeof(entry->update_stat));
+		entry->update_stat.st_min = ULONG_MAX;
+		memset(&entry->fetch_stat, 0, sizeof(entry->fetch_stat));
+		entry->fetch_stat.st_min = ULONG_MAX;
+		memset(&entry->ids, 0, sizeof(entry->ids));
+		memset(&entry->idp, 0, sizeof(entry->idp));
+	}
+	D_MUTEX_UNLOCK(&metrics_tls_lock);
+}
+
+static inline void
+dc_metrics_aggr_stats(daos_metrics_stat_t *dest, daos_metrics_stat_t *src)
+{
+	dest->st_value += src->st_value;
+	if (dest->st_min > src->st_min) {
+		dest->st_min = src->st_min;
+	}
+	if (dest->st_max < src->st_max) {
+		dest->st_max = src->st_max;
+	}
+	dest->st_sum += src->st_sum;
+	dest->st_sum_of_squares += src->st_sum_of_squares;
+}
+
+static void
+dc_metrics_tls_aggr(dc_metrics_tls_data_t *value)
+{
+	int i;
+
+	/** Update stats */
+	dc_metrics_aggr_stats(&metrics_tls_agg.update_stat, \
+			&value->update_stat);
+
+	/** Fetch stats */
+	dc_metrics_aggr_stats(&metrics_tls_agg.fetch_stat, \
+			&value->fetch_stat);
+
+	/** io distribution by size */
+	metrics_tls_agg.ids.ids_updatesz += value->ids.ids_updatesz;
+	metrics_tls_agg.ids.ids_fetchsz += value->ids.ids_fetchsz;
+	for (i=0; i< DAOS_METRICS_IO_BYSIZE_COUNT; i++) {
```

Comment:
(style)  spaces required around that '=' (ctx:VxV)
(style)  spaces required around that '<' (ctx:VxW)